Fix issue of invalid null-null-deltas due to resource-creation, build-cancellation, and resource-deletion

The following stack trace happens when a resource is 
(1) created,
(2) a cancellation happens, and
(3) the resource is removed again.
What should not happen is that the build crashes during this course of actions. A crash could happen since the delta that is created at (1) will be merged with the delta at (3). This merge will result in a delta from old==null to new==null, which is an invalid delta.

```
ERROR - error during queued task: 
java.lang.AssertionError: 'old!=_new' constraint violated
	at org.eclipse.xtext.resource.impl.DefaultResourceDescriptionDelta.<init>(DefaultResourceDescriptionDelta.java:30)
	at org.eclipse.n4js.ide.xtext.server.build.XWorkspaceBuilder.mergeWithUnreportedDeltas(XWorkspaceBuilder.java:438)
	at org.eclipse.n4js.ide.xtext.server.build.XWorkspaceBuilder.recordBuildProgress(XWorkspaceBuilder.java:420)
	at org.eclipse.n4js.ide.xtext.server.build.XWorkspaceBuilder.doIncrementalBuild(XWorkspaceBuilder.java:362)
	at org.eclipse.n4js.ide.xtext.server.build.XWorkspaceBuilder.lambda$5(XWorkspaceBuilder.java:281)
	at org.eclipse.n4js.ide.xtext.server.build.BuilderFrontend.lambda$5(BuilderFrontend.java:157)
	at org.eclipse.n4js.ide.server.concurrent.N4JSQueuedExecutorService.lambda$0(N4JSQueuedExecutorService.java:35)
	at org.eclipse.n4js.ide.xtext.server.QueuedExecutorService$QueuedTask.run(QueuedExecutorService.java:119)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

--- a/plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XBuildRequest.java
+++ b/plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XBuildRequest.java
@@ -10,6 +10,7 @@ package org.eclipse.n4js.ide.xtext.server.build;
 import java.io.File;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 
@@ -33,11 +34,11 @@ public class XBuildRequest {
 
 	private URI baseDir;
 
-	private Collection<URI> dirtyFiles = new ArrayList<>();
+	private Collection<URI> dirtyFiles = Collections.emptyList();
 
-	private Collection<URI> deletedFiles = new ArrayList<>();
+	private Collection<URI> deletedFiles = Collections.emptyList();
 
-	private Collection<IResourceDescription.Delta> externalDeltas = new ArrayList<>();
+	private Collection<IResourceDescription.Delta> externalDeltas = Collections.emptyList();
 
 	private ResourceDescriptionsData index;
 
@@ -126,7 +127,7 @@ public class XBuildRequest {
 
 	/** Setter. */
 	public void setDirtyFiles(Collection<URI> dirtyFiles) {
-		this.dirtyFiles = dirtyFiles;
+		this.dirtyFiles = Collections.unmodifiableCollection(dirtyFiles);
 	}
 
 	/** Getter. */
@@ -136,7 +137,7 @@ public class XBuildRequest {
 
 	/** Setter. */
 	public void setDeletedFiles(Collection<URI> deletedFiles) {
-		this.deletedFiles = deletedFiles;
+		this.deletedFiles = Collections.unmodifiableCollection(deletedFiles);
 	}
 
 	/** Getter. */
@@ -146,7 +147,7 @@ public class XBuildRequest {
 
 	/** Setter. */
 	public void setExternalDeltas(Collection<IResourceDescription.Delta> externalDeltas) {
-		this.externalDeltas = externalDeltas;
+		this.externalDeltas = Collections.unmodifiableCollection(externalDeltas);
 	}
 
 	/**

--- a/plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceBuilder.java
+++ b/plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceBuilder.java
@@ -435,7 +435,9 @@ public class XWorkspaceBuilder {
 					toBeConsideredDeltas.remove(unreportedDelta);
 					IResourceDescription _old = unreportedDelta.getOld();
 					IResourceDescription _new = newDelta.getNew();
-					toBeConsideredDeltas.add(new DefaultResourceDescriptionDelta(_old, _new));
+					if (_old != null || _new != null) {
+						toBeConsideredDeltas.add(new DefaultResourceDescriptionDelta(_old, _new));
+					}
 				}
 			}
 		}

--- a/plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceBuilder.java
+++ b/plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/build/XWorkspaceBuilder.java
@@ -435,9 +435,13 @@ public class XWorkspaceBuilder {
 					toBeConsideredDeltas.remove(unreportedDelta);
 					IResourceDescription _old = unreportedDelta.getOld();
 					IResourceDescription _new = newDelta.getNew();
-					if (_old != null || _new != null) {
-						toBeConsideredDeltas.add(new DefaultResourceDescriptionDelta(_old, _new));
+					if (_old == null && _new == null) {
+						// happens in case a resource was created, the build was cancelled and deleted again
+						// before the next build
+						_old = newDelta.getOld();
 					}
+
+					toBeConsideredDeltas.add(new DefaultResourceDescriptionDelta(_old, _new));
 				}
 			}
 		}

--- a/tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/builder/IncrementalBuilderCancellationTest.xtend
+++ b/tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/builder/IncrementalBuilderCancellationTest.xtend
@@ -27,17 +27,22 @@ import org.eclipse.xtext.util.Strings
 import org.junit.Test
 
 import static org.junit.Assert.*
+import java.util.concurrent.atomic.AtomicReference
+import org.eclipse.n4js.ide.tests.server.TestWorkspaceManager
 
 /**
  * Tests handling of a cancellation during an incremental build.
  */
 class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest {
 
+	private static final AtomicReference<URI> cancelOnResource = new AtomicReference(null);
+	private static final AtomicReference<URI> cancelAfterResource = new AtomicReference(null);
+	
 	/* 
 	 * Iff greater zero, then cancel before building the n-th resource. Cancellation happens before the very first 
 	 * resource with a cancelCounter set to 1.
 	 */
-	private static final AtomicInteger cancelCounter = new AtomicInteger(0);
+	private static final AtomicInteger cancelOnCount = new AtomicInteger(0);
 
 	private static final class CancelingStatefulIncrementalBuilder extends XStatefulIncrementalBuilder {
 
@@ -45,13 +50,25 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 		private OperationCanceledManager operationCanceledManager;
 		
 		override protected buildClustered(LoadResult loadResult, XSource2GeneratedMapping newSource2GeneratedMapping, XIndexResult result) {
-			var isCancelRequested = false;
-			synchronized(cancelCounter) {
-				isCancelRequested = cancelCounter.get() > 0 && cancelCounter.decrementAndGet() == 0;
+			var isCancelledByCounter = false;
+			var isCancelledByResource = false;
+
+			synchronized(cancelOnCount) {
+				isCancelledByCounter = cancelOnCount.get() > 0 && cancelOnCount.decrementAndGet() == 0;
 			}
-			if (isCancelRequested) {
+			synchronized(cancelOnResource) {
+				isCancelledByResource = cancelOnResource.get() !== null && cancelOnResource.get() == loadResult.uri;
+			}
+			synchronized(cancelAfterResource) {
+				if (cancelAfterResource.get() !== null && cancelAfterResource.get() == loadResult.uri) {
+					cancelOnCount.set(1);
+				}
+			}
+
+			if (isCancelledByResource || isCancelledByCounter) {
 				operationCanceledManager.checkCanceled([true]);
 			}
+			
 			super.buildClustered(loadResult, newSource2GeneratedMapping, result);
 		}
 	}
@@ -106,10 +123,10 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 		joinServerRequests();
 		assertNoIssues();
 
-		cancelCounter.set(4);
+		cancelOnCount.set(4);
 		saveOpenedFile("Main");
 		joinServerRequests();
-		assertEquals(0, cancelCounter.get());
+		assertEquals(0, cancelOnCount.get());
 
 		// With a cancelCounter of 4 we expect a cancellation to occur in the fourth file being built.
 		// Since "Main" is built first and counted as well, we expect the cancellation to occur in the
@@ -155,10 +172,10 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 			issues.sort.join(Strings.newLine))
 
 		// cancel before A is validated - no error changed
-		cancelCounter.set(1);
+		cancelOnCount.set(1);
 		changeNonOpenedFile("Main", "methA(" -> "methB(");
 		joinServerRequests();
-		assertEquals(0, cancelCounter.get());
+		assertEquals(0, cancelOnCount.get());
 		issues = getIssues().values().map[getStringLSP4J.toStringShort(it)].toSet;
 		assertEquals(3, issues.size);
 		assertEquals('''
@@ -205,10 +222,10 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 			issues.sort.join(Strings.newLine))
 
 		// Rename a2 to a3 to fix the reference in C but cancel, after C was processed
-		cancelCounter.set(4);
+		cancelOnCount.set(4);
 		changeNonOpenedFile("A", "a1 : { a2 : '' }" -> "a0 : {}, a1 : { a3 : '' }");
 		joinServerRequests();
-		assertEquals(0, cancelCounter.get());
+		assertEquals(0, cancelOnCount.get());
 		issues = getIssues().values().map[getStringLSP4J.toStringShort(it)].toSet;
 		assertEquals(issues.sort.join(Strings.newLine), 0, issues.size);
 			
@@ -218,5 +235,100 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 		issues = getIssues().values().map[getStringLSP4J.toStringShort(it)].toSet;
 		assertEquals("(Error, [1:24 - 1:26], Couldn't resolve reference to IdentifiableElement 'a3'.)",
 			issues.sort.join(Strings.newLine))
+	}
+
+	/**
+	 * This test case tests a very specific edge case: A resource is 
+	 * (1) created,
+	 * (2) a cancellation happens, and
+	 * (3) the resource is removed again.
+	 * What should not happen is that the build crashes during this course of actions.
+	 * A crash could happen since the delta that is created at (1) will be merged with the delta at (3).
+	 * This merge will result in a delta from old==null to new==null, which is an invalid delta.
+	 */
+	@Test
+	def void testDeletionDuringCancelledBuild() {
+		testWorkspaceManager.createTestOnDisk(
+			TestWorkspaceManager.CFG_NODE_MODULES + "n4js-runtime" -> null,
+			"ProjectClientA" -> #[
+				"MainClientA" -> '''
+					import {Cls} from "MainLibTEMP";
+					new Cls().meth();
+					export public class ClsA {
+						public methA() {}
+					}
+				''',
+				TestWorkspaceManager.CFG_DEPENDENCIES -> '''
+					n4js-runtime,
+					ProjectLib
+				'''
+			],
+			"ProjectClientB" -> #[
+				"MainClientB" -> '''
+					import {ClsA} from "MainClientA";
+					new ClsA().methA();
+				''',
+				TestWorkspaceManager.CFG_DEPENDENCIES -> '''
+					n4js-runtime,
+					ProjectClientA
+				'''
+			],
+			"ProjectLib" -> #[
+				"MainLib" -> '''
+					export public class Cls {
+						public meth() {}
+					}
+				''',
+				TestWorkspaceManager.CFG_DEPENDENCIES -> '''
+					n4js-runtime
+				'''
+			]
+		);
+		
+		
+		val uriMainClient = getFileURIFromModuleName("MainClientA");
+		startAndWaitForLspServer();
+		assertIssues(
+			"MainClientA" -> #[
+				"(Error, [0:18 - 0:31], Cannot resolve plain module specifier (without project name as first segment): no matching module found.)",
+				"(Error, [1:4 - 1:7], Couldn't resolve reference to IdentifiableElement 'Cls'.)"]);
+		
+		cancelOnResource.set(uriMainClient.toURI());
+		createFileOnDiskWithoutNotification("MainLib", "MainLibTEMP.n4js", '''
+				export public class Cls {
+					public meth() {}
+				}
+			''');
+		// this change should fix the issue
+		changeNonOpenedFile("MainLibTEMP", "public meth() {}" -> '''
+				public meth() {}
+				public methNew() {} // trigger rebuild of MainClient
+			''');
+		joinServerRequests();
+		
+		// due to cancellation the errors still remain
+		assertIssues(
+			"MainClientA" -> #[
+				"(Error, [0:18 - 0:31], Cannot resolve plain module specifier (without project name as first segment): no matching module found.)",
+				"(Error, [1:4 - 1:7], Couldn't resolve reference to IdentifiableElement 'Cls'.)"]);
+		
+		
+		val uriMainLib_TEMP = getFileURIFromModuleName("MainLibTEMP");
+		cancelOnResource.set(null);
+		// perform two different changes
+		// Change #1: introduce new errors that should show up
+		changeFileOnDiskWithoutNotification("MainClientA", "public methA() {}" -> "public methXXX() {} // introduce error");
+		// Change #2: delete file to cause the error to occur
+		deleteNonOpenedFile(uriMainLib_TEMP);
+		joinServerRequests();
+		
+		
+		assertIssues(
+			"MainClientA" -> #[
+				"(Error, [0:18 - 0:31], Cannot resolve plain module specifier (without project name as first segment): no matching module found.)",
+				"(Error, [1:4 - 1:7], Couldn't resolve reference to IdentifiableElement 'Cls'.)"],
+			// This error is due to change #2
+			"MainClientB" -> #[
+				"(Error, [1:11 - 1:16], Couldn't resolve reference to IdentifiableElement 'methA'.)"]);
 	}
 }

--- a/tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/builder/IncrementalBuilderCancellationTest.xtend
+++ b/tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/builder/IncrementalBuilderCancellationTest.xtend
@@ -38,11 +38,11 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 	private static final AtomicReference<URI> cancelOnResource = new AtomicReference(null);
 	private static final AtomicReference<URI> cancelAfterResource = new AtomicReference(null);
 	
-	/* 
+	/**
 	 * Iff greater zero, then cancel before building the n-th resource. Cancellation happens before the very first 
 	 * resource with a cancelCounter set to 1.
 	 */
-	private static final AtomicInteger cancelOnCount = new AtomicInteger(0);
+	private static final AtomicInteger cancelOnCountZero = new AtomicInteger(0);
 
 	private static final class CancelingStatefulIncrementalBuilder extends XStatefulIncrementalBuilder {
 
@@ -53,15 +53,15 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 			var isCancelledByCounter = false;
 			var isCancelledByResource = false;
 
-			synchronized(cancelOnCount) {
-				isCancelledByCounter = cancelOnCount.get() > 0 && cancelOnCount.decrementAndGet() == 0;
+			synchronized(cancelOnCountZero) {
+				isCancelledByCounter = cancelOnCountZero.get() > 0 && cancelOnCountZero.decrementAndGet() == 0;
 			}
 			synchronized(cancelOnResource) {
 				isCancelledByResource = cancelOnResource.get() !== null && cancelOnResource.get() == loadResult.uri;
 			}
 			synchronized(cancelAfterResource) {
 				if (cancelAfterResource.get() !== null && cancelAfterResource.get() == loadResult.uri) {
-					cancelOnCount.set(1);
+					cancelOnCountZero.set(1);
 				}
 			}
 
@@ -123,10 +123,10 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 		joinServerRequests();
 		assertNoIssues();
 
-		cancelOnCount.set(4);
+		cancelOnCountZero.set(4);
 		saveOpenedFile("Main");
 		joinServerRequests();
-		assertEquals(0, cancelOnCount.get());
+		assertEquals(0, cancelOnCountZero.get());
 
 		// With a cancelCounter of 4 we expect a cancellation to occur in the fourth file being built.
 		// Since "Main" is built first and counted as well, we expect the cancellation to occur in the
@@ -172,10 +172,10 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 			issues.sort.join(Strings.newLine))
 
 		// cancel before A is validated - no error changed
-		cancelOnCount.set(1);
+		cancelOnCountZero.set(1);
 		changeNonOpenedFile("Main", "methA(" -> "methB(");
 		joinServerRequests();
-		assertEquals(0, cancelOnCount.get());
+		assertEquals(0, cancelOnCountZero.get());
 		issues = getIssues().values().map[getStringLSP4J.toStringShort(it)].toSet;
 		assertEquals(3, issues.size);
 		assertEquals('''
@@ -222,10 +222,10 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 			issues.sort.join(Strings.newLine))
 
 		// Rename a2 to a3 to fix the reference in C but cancel, after C was processed
-		cancelOnCount.set(4);
+		cancelOnCountZero.set(4);
 		changeNonOpenedFile("A", "a1 : { a2 : '' }" -> "a0 : {}, a1 : { a3 : '' }");
 		joinServerRequests();
-		assertEquals(0, cancelOnCount.get());
+		assertEquals(0, cancelOnCountZero.get());
 		issues = getIssues().values().map[getStringLSP4J.toStringShort(it)].toSet;
 		assertEquals(issues.sort.join(Strings.newLine), 0, issues.size);
 			
@@ -241,7 +241,7 @@ class IncrementalBuilderCancellationTest extends AbstractIncrementalBuilderTest 
 	 * This test case tests a very specific edge case: A resource is 
 	 * (1) created,
 	 * (2) a cancellation happens, and
-	 * (3) the resource is removed again.
+	 * (3) the resource is removed again before the next incremental build iteration.
 	 * What should not happen is that the build crashes during this course of actions.
 	 * A crash could happen since the delta that is created at (1) will be merged with the delta at (3).
 	 * This merge will result in a delta from old==null to new==null, which is an invalid delta.

--- a/tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/server/AbstractIdeTest.java
+++ b/tests/org.eclipse.n4js.ide.tests/src/org/eclipse/n4js/ide/tests/server/AbstractIdeTest.java
@@ -12,6 +12,7 @@ package org.eclipse.n4js.ide.tests.server;
 
 import static java.util.Collections.singletonList;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
 
 import java.io.File;
 import java.io.IOException;
@@ -54,6 +55,7 @@ import org.eclipse.lsp4j.FileChangeType;
 import org.eclipse.lsp4j.FileEvent;
 import org.eclipse.lsp4j.InitializeParams;
 import org.eclipse.lsp4j.MessageParams;
+import org.eclipse.lsp4j.MessageType;
 import org.eclipse.lsp4j.Range;
 import org.eclipse.lsp4j.RenameCapabilities;
 import org.eclipse.lsp4j.ResourceChange;
@@ -95,6 +97,7 @@ import org.junit.BeforeClass;
 
 import com.google.common.base.Joiner;
 import com.google.common.base.Optional;
+import com.google.common.base.Strings;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Lists;
@@ -169,6 +172,7 @@ abstract public class AbstractIdeTest implements IIdeTestLanguageClientListener 
 	/** Deletes the test project in case it exists. */
 	@After
 	final public void deleteTestProject() {
+		assertNoErrorsInLog();
 		shutdownLspServer();
 		// clear the state related to the test
 		testWorkspaceManager.deleteTestFromDiskIfCreated();
@@ -894,6 +898,14 @@ abstract public class AbstractIdeTest implements IIdeTestLanguageClientListener 
 	 */
 	protected void assertNoIssues() {
 		assertIssues(Collections.emptyMap());
+	}
+
+	/** Asserts that there are no ERRORs in the LSP log. */
+	protected void assertNoErrorsInLog() {
+		for (MessageParams msg : languageClient.getLogMessages()) {
+			String message = Strings.nullToEmpty(msg.getMessage());
+			assertNotEquals("Unexpected ERROR in log:\n" + message, MessageType.Error, msg.getType());
+		}
 	}
 
 	/**

